connectors/sources/sharepoint.py

```diff
@@ -0,0 +1,715 @@
+            },
+            "ssl_disabled": {
+                "value": True,
+                "label": "Disable SSL verification",
+                "type": "bool",
+            },
+            "ssl_ca": {
+                "value": "",
+                "label": "SSL certificate",
+                "type": "str",
+            },
+            "enable_content_extraction": {
+                "value": True,
+                "label": "Enable content extraction (true/false)",
+                "type": "bool",
+            },
+            "retry_count": {
+                "value": RETRIES,
+                "label": "Maximum retries per request",
+                "type": "int",
+            },
+        }
+
+    def _validate_configuration(self):
+        """Validates whether user input is empty or not for configuration fields
+
+        Raises:
+            Exception: Configured keys can't be empty.
+        """
+        logger.info("Validating Sharepoint Configuration...")
+
+        connection_fields = (
+            [
+                "host_url",
+                "site_collections",
+                "client_id",
+                "secret_id",
+                "tenant",
+                "tenant_id",
+            ]
+            if self.is_cloud
+            else ["host_url", "site_collections", "username", "password"]
+        )
+
+        if empty_connection_fields := [
+            field for field in connection_fields if self.configuration[field] == ""
+        ]:
+            raise Exception(
+                f"Configured keys: {empty_connection_fields} can't be empty."
+            )
+        if not self.ssl_disabled and self.certificate == "":
+            raise Exception("Configured key ssl_ca can't be empty when SSL is enabled")
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def _set_access_token(self):
+        """Set access token using configuration fields"""
+        if not is_expired(expires_at=self.token_expires_in):
+            return
+        tenant_id = self.configuration["tenant_id"]
+        logger.info("Generating access token...")
+        url = f"https://accounts.accesscontrol.windows.net/{tenant_id}/tokens/OAuth/2"
+        data = {
+            "grant_type": "client_credentials",
+            "resource": f"00000003-0000-0ff1-ce00-000000000000/{self.configuration['tenant']}.sharepoint.com@{tenant_id}",
+            "client_id": f"{self.configuration['client_id']}@{tenant_id}",
+            "client_secret": self.configuration["secret_id"],
+        }
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+
+        async with aiohttp.request(
+            method="POST", url=url, data=data, headers=headers
+        ) as response:
+            json_data = await response.json()
+            self.access_token = json_data["access_token"]
+            self.token_expires_in = get_expires_at(int(json_data["expires_in"]))
```

Comment:
We are setting the value of `token_expires_in` here because every token has a expiry time which needs to replaced every time token expires with the newer expiry time.

Reply:
we are planning to add a client call which will take care of token generation as well as the API call just like we did in confluence. but since all the testing are carried on this architecture, it will be better if we can merge this PR meanwhile we have already started that enhancement and we would be raising and incremental the PR once this gets merged.